core/src/main/java/org/apache/iceberg/encryption/StandardEncryptionManagerFactory.java

```diff
@@ -0,0 +1,110 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import static org.apache.iceberg.TableProperties.DEFAULT_FILE_FORMAT;
+import static org.apache.iceberg.TableProperties.DEFAULT_FILE_FORMAT_DEFAULT;
+
+import java.io.IOException;
+import java.util.Map;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.PropertyUtil;
+
+public class StandardEncryptionManagerFactory implements EncryptionManagerFactory {
+  private KeyManagementClient kmsClient;
+  private Map<String, String> catalogPropertyMap;
+
+  @Override
+  public void initialize(Map<String, String> catalogProperties) {
+    this.catalogPropertyMap = catalogProperties;
+  }
+
+  @Override
+  public EncryptionManager create(TableMetadata tableMetadata) {
+    if (tableMetadata == null) {
+      return PlaintextEncryptionManager.instance();
+    }
+
+    Map<String, String> tableProperties = tableMetadata.properties();
+
+    final Map<String, String> encryptionProperties = Maps.newHashMap();
+    encryptionProperties.putAll(tableProperties);
+
+    // Important: put catalog properties after table properties. Former overrides the latter.
+    encryptionProperties.putAll(catalogPropertyMap);
+
+    String tableKeyId = encryptionProperties.get(EncryptionProperties.ENCRYPTION_TABLE_KEY);
+
+    if (null == tableKeyId) {
+      // Unencrypted table
+      return PlaintextEncryptionManager.instance();
+    } else {
+      String fileFormat =
+          PropertyUtil.propertyAsString(
+              tableProperties, DEFAULT_FILE_FORMAT, DEFAULT_FILE_FORMAT_DEFAULT);
+
+      if (FileFormat.fromString(fileFormat) != FileFormat.PARQUET) {
+        throw new UnsupportedOperationException(
+            "Iceberg encryption currently supports only parquet format for data files");
+      }
+
+      if (tableMetadata.formatVersion() < 2) {
+        throw new UnsupportedOperationException(
+            "Iceberg encryption works only with table format 2 or higher");
+      }
+
+      int dataKeyLength =
+          PropertyUtil.propertyAsInt(
+              encryptionProperties,
+              EncryptionProperties.ENCRYPTION_DEK_LENGTH,
+              EncryptionProperties.ENCRYPTION_DEK_LENGTH_DEFAULT);
+
+      return new StandardEncryptionManager(
+          tableKeyId, dataKeyLength, kmsClient(encryptionProperties), encryptionProperties);
+    }
+  }
+
+  private synchronized KeyManagementClient kmsClient(Map<String, String> encryptionProperties) {
+    if (kmsClient == null) {
```

Comment:
If `encryptionProperties` can be different for every table, then this cannot cache the KMS client. The impl or other properties may change.

I think the intent is to use one KMS for all tables in a given catalog. If that's the case, then this should only use catalog properties and it should be initialized in the `initialize` method or the constructor.

Reply:
Following our discussion, moving the multi-kms logic into kms client implementations.